haskell/import.bzl

```diff
@@ -0,0 +1,104 @@
+"""Importing prebuilt packages into bazel"""
+
+load(":private/context.bzl", "haskell_context")
+load(":private/actions/package.bzl", "package")
+load(
+    ":private/providers.bzl",
+    "HaddockInfo",
+    "HaskellBuildInfo",
+    "HaskellLibraryInfo",
+)
+load(":private/set.bzl", "set")
+load(":private/path_utils.bzl", "copy_all", "link_forest", "ln")
+load("@bazel_skylib//:lib.bzl", "paths")
+
+def _haskell_import_impl(ctx):
+    hs = haskell_context(ctx)
+
+    package_cache = ctx.actions.declare_file(
+        paths.join("package.conf.d", "package.cache"),
+    )
+
+    local_package_confs = link_forest(
+        ctx = ctx,
+        srcs = ctx.attr.package_confs.files,
+        sibling = package_cache,
+    )
+
+    local_haddock_html = ctx.actions.declare_directory("haddock-html")
+    copy_all(
+        ctx = ctx,
+        srcs = ctx.attr.haddock_html.files,
+        dest = local_haddock_html,
+    )
+
+    ctx.actions.run(
+        outputs = [package_cache],
+        inputs = local_package_confs,
+        mnemonic = "HaskellCreatePackageCache",
+        executable = hs.tools.ghc_pkg,
+        arguments = [
+            "recache",
+            "--package-db",
+            package_cache.dirname,
+        ],
+    )
+    ln(ctx, package_cache, ctx.outputs.cache)
+
+    dependencies_caches = set.singleton(package_cache)
+    for dep in ctx.attr.deps:
+        if HaskellBuildInfo in dep:
+            set.mutable_union(dependencies_caches, dep[HaskellBuildInfo].package_caches)
+
+    libInfo = HaskellLibraryInfo(
+        package_id = ctx.attr.package_id,
+        version = ctx.attr.version,
+        import_dirs = [],
+        header_files = set.empty(),
+        boot_files = set.empty(),
+        source_files = set.empty(),
+        extra_source_files = set.empty(),
+        ghc_args = [],
+    )
+    buildInfo = HaskellBuildInfo(
+        package_ids = set.from_list([ctx.attr.package_id] + ctx.attr.deps_ids),
+        package_confs = set.from_list(local_package_confs),
+        package_caches = dependencies_caches,
+        static_libraries = [],
+        static_libraries_prof = [],
+        dynamic_libraries = set.from_list(ctx.attr.dynamic_libraries),
+        interface_dirs = set.empty(),
+        prebuilt_dependencies = set.empty(),
+        external_libraries = {},
+        direct_prebuilt_deps = set.empty(),
+    )
+    html_files = list(ctx.attr.haddock_html.files)
+    transitive_html = {ctx.attr.package_id: local_haddock_html}
+    interface_files = list(ctx.attr.haddock_interfaces.files)
+    transitive_haddocks = {ctx.attr.package_id: interface_files[0]} if interface_files != [] else {}
```

Comment:
Adding an explanatory comment to this line is the last thing from my side, then +1 for the merge.